### PR TITLE
Initial Object Store implementation

### PR DESCRIPTION
**Pull request by mtmk**
Object store initial implementation.

Features:
* Create/delete stores
* get/put/delete objects

Missing features:
* linked objects/buckets

### CODE REVIEW

File: src/NATS.Client.Core/INatsSerializer.cs [hunk 1]
```diff
@@ -18,6 +18,21 @@ public interface ICountableBufferWriter : IBufferWriter<byte>
     int WrittenCount { get; }
 }
 
+public readonly struct FixedSizeMemoryOwner : IMemoryOwner<byte>
```
Comment (caleblloyd) [inline src/NATS.Client.Core/INatsSerializer.cs:21]: Out of curiosity, what is the benefit of using this over `MemoryPool<byte>.Shared` ?

Comment (mtmk) [reply src/NATS.Client.Core/INatsSerializer.cs:21]: MemoryPool<byte>.Shared.Rent returns *at least* the size asked for, which usually rounds up to some 2^n. This wrapper chops the memory to exact size so when it's passed to stream read/write or down the line to publish or something you won't need to pass the actual size.

Comment (mtmk) [reply src/NATS.Client.Core/INatsSerializer.cs:21]: btw, we are still renting it from MemoryPool.

Comment (caleblloyd) [reply src/NATS.Client.Core/INatsSerializer.cs:21]: > btw, we are still renting it from MemoryPool.

Ah, I see that now.  Clever!

File: src/NATS.Client.ObjectStore/NatsOBStore.cs [hunk 2]
```diff
@@ -0,0 +1,374 @@
...
+
+    /// <summary>
+    /// Get object by key.
+    /// </summary>
+    /// <param name="key">Object key.</param>
+    /// <param name="stream">Stream to write the object value to.</param>
+    /// <param name="cancellationToken">A <see cref="CancellationToken"/> used to cancel the API call.</param>
+    /// <returns>Object metadata.</returns>
+    /// <exception cref="NatsOBException">Metadata didn't match the value retrieved e.g. the SHA digest.</exception>
+    public async ValueTask<ObjectMetadata> GetAsync(string key, Stream stream, CancellationToken cancellationToken = default)
```
[the remark this answers is not in the document]
Comment (mtmk) [reply src/NATS.Client.ObjectStore/NatsOBStore.cs]: fixed

File: src/NATS.Client.ObjectStore/NatsOBStore.cs [hunk 3]
```diff
@@ -0,0 +1,374 @@
...
+    /// <summary>
+    /// Put an object by key.
+    /// </summary>
+    /// <param name="key">Object key.</param>
+    /// <param name="stream">Stream to read the value from.</param>
+    /// <param name="cancellationToken">A <see cref="CancellationToken"/> used to cancel the API call.</param>
+    /// <returns>Object metadata.</returns>
+    /// <exception cref="NatsOBException">There was an error calculating SHA digest.</exception>
+    /// <exception cref="NatsJSApiException">Server responded with an error.</exception>
+    public ValueTask<ObjectMetadata> PutAsync(string key, Stream stream, CancellationToken cancellationToken = default) =>
```
[the remark this answers is not in the document]
Comment (mtmk) [reply src/NATS.Client.ObjectStore/NatsOBStore.cs:153]: fixed

File: src/NATS.Client.ObjectStore/NatsOBStore.cs [hunk 4]
```diff
@@ -150,18 +151,19 @@ await foreach (var msg in pushConsumer.Msgs.ReadAllAsync(cancellationToken))
     /// <exception cref="NatsOBException">There was an error calculating SHA digest.</exception>
     /// <exception cref="NatsJSApiException">Server responded with an error.</exception>
     public ValueTask<ObjectMetadata> PutAsync(string key, Stream stream, CancellationToken cancellationToken = default) =>
-        PutAsync(new ObjectMetadata { Name = key }, stream, cancellationToken);
+        PutAsync(new ObjectMetadata { Name = key }, stream, cancellationToken: cancellationToken);
```
Comment (caleblloyd) [inline src/NATS.Client.ObjectStore/NatsOBStore.cs:154]: this one still needs `leaveOpen`

Comment (mtmk) [reply src/NATS.Client.ObjectStore/NatsOBStore.cs:154]: ah! missed that

Comment (mtmk) [reply src/NATS.Client.ObjectStore/NatsOBStore.cs]: all done 🤞

**Review by caleblloyd — APPROVED** [review]
LGTM

**Pull request merged**